File: backend/app/schemas/slo.py
"""ARCH-17 — SLO API contracts."""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.slo import SLOMethod, SLOUnit, SLOWindow


class SLOTarget(BaseModel):
    slo_key: str
    display_name: str
    description: str
    unit: SLOUnit
    target_value: Decimal
    window_period: SLOWindow
    is_contractual: bool
    source: str
    definition_id: Optional[uuid.UUID] = None
    notes: Optional[str] = None

    model_config = ConfigDict(protected_namespaces=())


class SLOComplianceEntry(BaseModel):
    slo_key: str
    target: SLOTarget

    observed_value: Optional[Decimal] = Field(
        default=None,
        description="None means no samples in the current window, not zero.",
    )
    sample_count: int
    error_count: int
    breached: bool
    method: Optional[SLOMethod] = Field(
        default=None,
        description=(
            "HISTOGRAM_INTERPOLATED means observed_value is an estimate whose "
            "error is bounded by one bucket width. The breach verdict is still "
            "exact when the target sits on a bucket boundary, which it always "
            "does for a contractual SLO."
        ),
    )

    window_start: datetime
    window_end: datetime

    breached_windows: int
    total_windows: int
    compliance_ratio: Optional[Decimal] = Field(
        default=None,
        description="Sealed windows met / sealed windows observed. None when none are sealed yet.",
    )

    model_config = ConfigDict(protected_namespaces=())


class SLOSummaryResponse(BaseModel):
    organization_id: uuid.UUID
    as_of: datetime
    period: SLOWindow
    contractual_breaches: int
    entries: list[SLOComplianceEntry]

    model_config = ConfigDict(protected_namespaces=())


class SLOTargetUpdate(BaseModel):
    target_value: Decimal = Field(ge=0)
    window_period: Optional[SLOWindow] = None
    is_contractual: bool = False
    notes: Optional[str] = Field(default=None, max_length=1000)

    @model_validator(mode="after")
    def sane_bounds(self):
        if self.target_value < 0:
            raise ValueError("A target cannot be negative.")
        return self

    model_config = ConfigDict(protected_namespaces=())


__all__ = [
    "SLOComplianceEntry",
    "SLOSummaryResponse",
    "SLOTarget",
    "SLOTargetUpdate",
]

File: frontend/src/types/slo.ts
/**
 * ARCH-17 — per-tenant SLO Data Transfer Objects.
 */

export type SLOUnit = "MILLISECONDS" | "RATIO";

export type SLOWindow = "HOUR" | "DAY" | "MONTH";

export type SLOMethod = "EXACT" | "HISTOGRAM_INTERPOLATED";

export type SLOSource = "ORGANIZATION" | "PLATFORM_DEFAULT" | "REGISTRY_DEFAULT";

export interface SLOTarget {
  readonly slo_key: string;
  readonly display_name: string;
  readonly description: string;
  readonly unit: SLOUnit;
  readonly target_value: string;
  readonly window_period: SLOWindow;
  readonly is_contractual: boolean;
  readonly source: SLOSource;
  readonly definition_id: string | null;
  readonly notes: string | null;
}

export interface SLOComplianceEntry {
  readonly slo_key: string;
  readonly target: SLOTarget;
  readonly observed_value: string | null;
  readonly sample_count: number;
  readonly error_count: number;
  readonly breached: boolean;
  readonly method: SLOMethod | null;
  readonly window_start: string;
  readonly window_end: string;
  readonly breached_windows: number;
  readonly total_windows: number;
  readonly compliance_ratio: string | null;
}

export interface SLOSummary {
  readonly organization_id: string;
  readonly as_of: string;
  readonly period: SLOWindow;
  readonly contractual_breaches: number;
  readonly entries: readonly SLOComplianceEntry[];
}

export interface SLOTargetUpdateRequest {
  readonly target_value: string;
  readonly window_period?: SLOWindow;
  readonly is_contractual: boolean;
  readonly notes?: string | null;
}

export const formatSLOValue = (
  value: string | null,
  unit: SLOUnit,
): string => {
  if (value === null) {return "—";}
  const numeric = Number(value);
  if (Number.isNaN(numeric)) {return "—";}
  if (unit === "RATIO") {return `${(numeric * 100).toFixed(2)}%`;}
  if (numeric >= 1000) {return `${(numeric / 1000).toFixed(2)}s`;}
  return `${Math.round(numeric)}ms`;
};

export const sloGaugeFraction = (entry: SLOComplianceEntry): number | null => {
  if (entry.observed_value === null) {return null;}
  const observed = Number(entry.observed_value);
  const target = Number(entry.target.target_value);
  if (!Number.isFinite(observed) || !Number.isFinite(target) || target === 0) {
    return null;
  }
  const fraction =
    entry.target.unit === "MILLISECONDS" ? observed / target : observed / target;
  return Math.max(0, Math.min(1, fraction));
};
